Implement PReLU in a compositional way

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #91238

The PReLU implementation was all over the place. This lead to a number
of bugs like https://github.com/pytorch/pytorch/issues/68760.  We fix it by:
- Keeping the weird broadcasting logic it has as a CompositeImplicit kernel that calls into a second kernel
- This second kernel is just a good-ol' pointwise kernel.
- We implement the derivative for the pointwise kernel via TI as well for speed.
- We implement the second derivative for the pointwise kernel and the forward AD derivatives compositionally

This fixes a number of issues:
- We don't perform copies any more when the inputs are not contiguous
- The derivatives are now correct
- We fix vmap and many other functorch-related issues.
- CPU and CUDA now share the relevant broadcasting logic
- The implementation is about 1/3 the length.


Fixes https://github.com/pytorch/pytorch/issues/68760
Fixes https://github.com/pytorch/pytorch/issues/89895

cc @albanD @mruberry @jbschlosser @walterddr @kshitij12345 @saketh-are @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

lezcano committed:
Implement PReLU in a compositional way

The PReLU implementation was all over the place. This lead to a number
of bugs like https://github.com/pytorch/pytorch/issues/68760. We port it
to TI and split it into broadcasting operations (CompositeImplicit) and
kernel fusion (via TI). This makes the CPU and CUDA parts agree and
heavily simplifies the code. This also includes a number of fixes like
using opmath_t consistently and  a more efficient backward implementation
in the non-contiguous case without copying the data.

Fixes https://github.com/pytorch/pytorch/issues/68760

[ghstack-poisoned]

lezcano commented:
@soulitzer @albanD you have been tagged as codeowners, but could you please review it?

The only issue with the current approach is that I was not able to port the MKLDNN and the MPS implementations to just implement the simpler new kernel, so those should still be buggy. If the relevant PoCs from Intel / Apple port these, we should be able to remove all the relevant entries from derivatives.yaml and native_functions.yaml

lezcano committed:
Update on "Implement PReLU in a compositional way"


The PReLU implementation was all over the place. This lead to a number
of bugs like https://github.com/pytorch/pytorch/issues/68760. We port it
to TI and split it into broadcasting operations (CompositeImplicit) and
kernel fusion (via TI). This makes the CPU and CUDA parts agree and
heavily simplifies the code. This also includes a number of fixes like
using opmath_t consistently and  a more efficient backward implementation
in the non-contiguous case without copying the data.

Fixes https://github.com/pytorch/pytorch/issues/68760

cc albanD mruberry jbschlosser walterddr kshitij12345 saketh-are jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

lezcano committed:
Update on "Implement PReLU in a compositional way"


The PReLU implementation was all over the place. This lead to a number
of bugs like https://github.com/pytorch/pytorch/issues/68760.  We fix it by:
- Keeping the weird broadcasting logic it has as a CompositeImplicit kernel that calls into a second kernel
- This second kernel is just a good-ol' pointwise kernel.
- We implement the derivative for the pointwise kernel via TI as well for speed.
- We implement the second derivative for the pointwise kernel and the forward AD derivatives compositionally

This fixes a number of issues:
- We don't perform copies any more when the inputs are not contiguous
- The derivatives are now correct
- We fix vmap
- CPU and CUDA now share the relevant broadcasting logic
- The implementation is about 1/3 the length.


Fixes https://github.com/pytorch/pytorch/issues/68760
Fixes https://github.com/pytorch/pytorch/issues/89895

cc albanD mruberry jbschlosser walterddr kshitij12345 saketh-are jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

lezcano committed:
Update on "Implement PReLU in a compositional way"


The PReLU implementation was all over the place. This lead to a number
of bugs like https://github.com/pytorch/pytorch/issues/68760.  We fix it by:
- Keeping the weird broadcasting logic it has as a CompositeImplicit kernel that calls into a second kernel
- This second kernel is just a good-ol' pointwise kernel.
- We implement the derivative for the pointwise kernel via TI as well for speed.
- We implement the second derivative for the pointwise kernel and the forward AD derivatives compositionally

This fixes a number of issues:
- We don't perform copies any more when the inputs are not contiguous
- The derivatives are now correct
- We fix vmap
- CPU and CUDA now share the relevant broadcasting logic
- The implementation is about 1/3 the length.


Fixes https://github.com/pytorch/pytorch/issues/68760
Fixes https://github.com/pytorch/pytorch/issues/89895

cc albanD mruberry jbschlosser walterddr kshitij12345 saketh-are jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

albanD commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -4649,25 +4649,34 @@
     MkldnnCPU: mkldnn_prelu
-    CPU: prelu_cpu
-    CUDA: prelu_cuda
     MPS: prelu_mps
```
cc @kulinseth if you are interested in this one.

albanD reviewed (commented):
I'm not sure why this has to be more complex than functions like amax which we can implement without a pair a functions?

kshitij12345 reviewed (approved):
LGTM (on functorch batching rule front) :)

ngimel commented on aten/src/ATen/native/Activation.cpp:
```diff
@@ -676,124 +679,60 @@ TORCH_IMPL_FUNC(threshold_backward_out)(const Tensor& grad, const Tensor& self,
-        at::detail::scalar_tensor_static(1, self.scalar_type(), kCPU);
-  }
+Tensor _prelu_kernel(const Tensor& self, const Tensor& weight) {
+  auto options = self.options().dtype(promoteTypes(self.scalar_type(), weight.scalar_type()));
```
I'm not sure we want type promoting behavior for `prelu`, and we don't have it currently.

ngimel commented on torch/_decomp/decompositions.py:
```diff
@@ -286,6 +286,22 @@ def prelu_backward(
 
 
+@register_decomposition(aten._prelu_kernel)
+@pw_cast_for_opmath
```
pw_cast_for_opmath not needed here, it only leads to more complicated IR

ngimel replied:
We are outputting either `self` or `weight * self`, in both cases there's no difference whether they are pre-converted to float or not here - `self` is a copy either way, `weight * self` would do whatever is needed in multiplication

lezcano replied:
I'll clean up all that code then.

ngimel commented on torch/_decomp/decompositions.py:
```diff
@@ -286,6 +286,22 @@ def prelu_backward(
+
+
+@register_decomposition(aten._prelu_kernel_backward)
+@pw_cast_for_opmath
```
similarly here

lezcano committed:
Update on "Implement PReLU in a compositional way"


The PReLU implementation was all over the place. This lead to a number
of bugs like https://github.com/pytorch/pytorch/issues/68760.  We fix it by:
- Keeping the weird broadcasting logic it has as a CompositeImplicit kernel that calls into a second kernel
- This second kernel is just a good-ol' pointwise kernel.
- We implement the derivative for the pointwise kernel via TI as well for speed.
- We implement the second derivative for the pointwise kernel and the forward AD derivatives compositionally

This fixes a number of issues:
- We don't perform copies any more when the inputs are not contiguous
- The derivatives are now correct
- We fix vmap and many other functorch-related issues.
- CPU and CUDA now share the relevant broadcasting logic
- The implementation is about 1/3 the length.


Fixes https://github.com/pytorch/pytorch/issues/68760
Fixes https://github.com/pytorch/pytorch/issues/89895

cc albanD mruberry jbschlosser walterddr kshitij12345 saketh-are jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

albanD commented:
I'm not sure why the fact that you have a custom backward changes this? Is it just because you want a simpler backward function?
But the backward kernel already doesn't need to handle broadcasting (which is handled in the engine).

If this implementation works, no need to re-do it again. Just wanted to know why we're going this way.

lezcano committed:
Update on "Implement PReLU in a compositional way"


The PReLU implementation was all over the place. This lead to a number
of bugs like https://github.com/pytorch/pytorch/issues/68760.  We fix it by:
- Keeping the weird broadcasting logic it has as a CompositeImplicit kernel that calls into a second kernel
- This second kernel is just a good-ol' pointwise kernel.
- We implement the derivative for the pointwise kernel via TI as well for speed.
- We implement the second derivative for the pointwise kernel and the forward AD derivatives compositionally

This fixes a number of issues:
- We don't perform copies any more when the inputs are not contiguous
- The derivatives are now correct
- We fix vmap and many other functorch-related issues.
- CPU and CUDA now share the relevant broadcasting logic
- The implementation is about 1/3 the length.


Fixes https://github.com/pytorch/pytorch/issues/68760
Fixes https://github.com/pytorch/pytorch/issues/89895

cc albanD mruberry jbschlosser walterddr kshitij12345 saketh-are jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

lezcano committed:
Update on "Implement PReLU in a compositional way"


The PReLU implementation was all over the place. This lead to a number
of bugs like https://github.com/pytorch/pytorch/issues/68760.  We fix it by:
- Keeping the weird broadcasting logic it has as a CompositeImplicit kernel that calls into a second kernel
- This second kernel is just a good-ol' pointwise kernel.
- We implement the derivative for the pointwise kernel via TI as well for speed.
- We implement the second derivative for the pointwise kernel and the forward AD derivatives compositionally

This fixes a number of issues:
- We don't perform copies any more when the inputs are not contiguous
- The derivatives are now correct
- We fix vmap and many other functorch-related issues.
- CPU and CUDA now share the relevant broadcasting logic
- The implementation is about 1/3 the length.


Fixes https://github.com/pytorch/pytorch/issues/68760
Fixes https://github.com/pytorch/pytorch/issues/89895

cc albanD mruberry jbschlosser walterddr kshitij12345 saketh-are jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jbschlosser commented:
> Quantized fails as well, no idea why :(

Registering `QuantizedCPU: prelu_quantized_cpu` underneath `_prelu_kernel` instead of `prelu` and removing the now-redundant weight-shaping logic fixes the tests for me:
https://github.com/pytorch/pytorch/blob/67c53d50e5b34a536cce2ad98b9ca5f65ff8a34d/aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp#L683-L690

But I'm not exactly clear on why the dispatcher is decomposing `prelu` -> `_prelu_kernel` instead of calling the registered `QuantizedCPU` in your current PR.

lezcano committed:
Update on "Implement PReLU in a compositional way"


The PReLU implementation was all over the place. This lead to a number
of bugs like https://github.com/pytorch/pytorch/issues/68760.  We fix it by:
- Keeping the weird broadcasting logic it has as a CompositeImplicit kernel that calls into a second kernel
- This second kernel is just a good-ol' pointwise kernel.
- We implement the derivative for the pointwise kernel via TI as well for speed.
- We implement the second derivative for the pointwise kernel and the forward AD derivatives compositionally

This fixes a number of issues:
- We don't perform copies any more when the inputs are not contiguous
- The derivatives are now correct
- We fix vmap and many other functorch-related issues.
- CPU and CUDA now share the relevant broadcasting logic
- The implementation is about 1/3 the length.


Fixes https://github.com/pytorch/pytorch/issues/68760
Fixes https://github.com/pytorch/pytorch/issues/89895

cc albanD mruberry jbschlosser walterddr kshitij12345 saketh-are jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

lezcano commented:
Thank you for the pointer @jbschlosser! I fixed that but now I'm getting an error, but I don't understand the quantised code so I don't really know what's going on there. Perhaps @Xia-Weiwen can have a look?

The MKLDNN part was written by @yanbing-j. Could you please tell me how could we fix that bit?

jbschlosser commented:
> I fixed that but now I'm getting an error, but I don't understand the quantised code so I don't really know what's going on there.

@lezcano what's the error you're getting?

lezcano committed:
Update on "Implement PReLU in a compositional way"


The PReLU implementation was all over the place. This lead to a number
of bugs like https://github.com/pytorch/pytorch/issues/68760.  We fix it by:
- Keeping the weird broadcasting logic it has as a CompositeImplicit kernel that calls into a second kernel
- This second kernel is just a good-ol' pointwise kernel.
- We implement the derivative for the pointwise kernel via TI as well for speed.
- We implement the second derivative for the pointwise kernel and the forward AD derivatives compositionally

This fixes a number of issues:
- We don't perform copies any more when the inputs are not contiguous
- The derivatives are now correct
- We fix vmap and many other functorch-related issues.
- CPU and CUDA now share the relevant broadcasting logic
- The implementation is about 1/3 the length.


Fixes https://github.com/pytorch/pytorch/issues/68760
Fixes https://github.com/pytorch/pytorch/issues/89895

cc albanD mruberry jbschlosser walterddr kshitij12345 saketh-are jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

lezcano commented:
The issue in quantized is the one that's in https://github.com/pytorch/pytorch/actions/runs/3795264825/jobs/6454831858 @Xia-Weiwen any idea what could be going on there?

Now I have pushed fixes for MKLDNN and (hopefully) MPS. The patch for MKLDNN needs the following patch in ideep
```git
diff --git a/include/ideep/operators/prelu.hpp b/include/ideep/operators/prelu.hpp
index f958c27..6591b1f 100644
--- a/include/ideep/operators/prelu.hpp
+++ b/include/ideep/operators/prelu.hpp
@@ -15,14 +15,6 @@ struct prelu_forward : public dnnl::prelu_forward {
     auto src_in = src;
     auto weight_in = weight;

-    // Reshape weight to src dimension
-    auto new_dims = src.get_dims();
-    if (src.ndims() != weight.ndims()) {
-      std::vector<dim> dim_w(src.ndims(), 1);
-      dim_w[1] = weight.get_dim(0);
-      weight_in.reshape(dim_w);
-    }
-
     auto src_desc = src_in.get_desc();
     auto weight_desc = weight_in.get_desc().to_format_any();

@@ -61,14 +53,6 @@ struct prelu_backward : public dnnl::prelu_backward {
     auto diff_dst_in = diff_dst;
     auto weight_dims = weight_in.get_dims();

-    // Reshape wieght to src dimension
-    auto new_dims = src.get_dims();
-    if (src.ndims() != weight.ndims()) {
-      std::vector<dim> dim_w(src.ndims(), 1);
-      dim_w[1] = weight.get_dim(0);
-      weight_in.reshape(dim_w);
-    }
-
     auto src_desc = src_in.get_desc();
     auto weight_desc = weight_in.get_desc().to_format_any();
     auto diff_dst_desc = diff_dst_in.get_desc();
@@ -105,11 +89,6 @@ struct prelu_backward : public dnnl::prelu_backward {
          {DNNL_ARG_DIFF_SRC, diff_src},
          {DNNL_ARG_DIFF_WEIGHTS, diff_weight},
          {DNNL_ARG_SCRATCHPAD, scratchpad}});
-
-    // Reshape weight back to original dimension
-    if (diff_weight.get_dims() != weight_dims) {
-      diff_weight.reshape(weight_dims);
-    }
   }
 };
```
@jgong5 what's the procedure to land this into ideep?

lezcano committed:
Update on "Implement PReLU in a compositional way"


The PReLU implementation was all over the place. This lead to a number
of bugs like https://github.com/pytorch/pytorch/issues/68760.  We fix it by:
- Keeping the weird broadcasting logic it has as a CompositeImplicit kernel that calls into a second kernel
- This second kernel is just a good-ol' pointwise kernel.
- We implement the derivative for the pointwise kernel via TI as well for speed.
- We implement the second derivative for the pointwise kernel and the forward AD derivatives compositionally

This fixes a number of issues:
- We don't perform copies any more when the inputs are not contiguous
- The derivatives are now correct
- We fix vmap and many other functorch-related issues.
- CPU and CUDA now share the relevant broadcasting logic
- The implementation is about 1/3 the length.


Fixes https://github.com/pytorch/pytorch/issues/68760
Fixes https://github.com/pytorch/pytorch/issues/89895

cc albanD mruberry jbschlosser walterddr kshitij12345 saketh-are jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jbschlosser commented:
@lezcano Whoops, the quantization issue is that `torch.ops.quantized.prelu` still points directly to the quantized kernel (which is now registered under `_prelu_kernel`) and it no longer has the weight reshaping logic, so the broadcasting is done incorrectly. That reshaping logic is only "redundant" when entering through `torch.ops.aten.prelu`, not when the composite part is bypassed. I think we can either:

* Add back in the weight reshaping logic for the quantization kernel, accepting redundancy when invoked through the composite `prelu` OR
* Point `torch.ops.quantized.prelu` -> `prelu` instead of the kernel registered under `_prelu_kernel`. I'd defer to a quantization expert as to how to do this / if this is a bad idea

yanbing-j commented:
Hi @lezcano ,

> Edit2. Note that the previous MKLDNN code was wrong in some cases if it supports 0-dim tensors. The code in core should fix that issue.

Can you explain more about the wrong cases? And why the weight reshape need to be removed, both in prelu C++ kernel and ideep? Thanks.

lezcano commented:
@yanbing-j I think the code would break in the case where input is 1D and weight is 0D, then the code `dim_w[1] = weight.get_dim(0);` would be executed and would break.
This code will not be necessary with this PR, as the inputs given to all the backends would already have the same number of dimensions.

lezcano committed:
Update on "Implement PReLU in a compositional way"


The PReLU implementation was all over the place. This lead to a number
of bugs like https://github.com/pytorch/pytorch/issues/68760.  We fix it by:
- Keeping the weird broadcasting logic it has as a CompositeImplicit kernel that calls into a second kernel
- This second kernel is just a good-ol' pointwise kernel.
- We implement the derivative for the pointwise kernel via TI as well for speed.
- We implement the second derivative for the pointwise kernel and the forward AD derivatives compositionally

This fixes a number of issues:
- We don't perform copies any more when the inputs are not contiguous
- The derivatives are now correct
- We fix vmap and many other functorch-related issues.
- CPU and CUDA now share the relevant broadcasting logic
- The implementation is about 1/3 the length.


Fixes https://github.com/pytorch/pytorch/issues/68760
Fixes https://github.com/pytorch/pytorch/issues/89895

cc albanD mruberry jbschlosser walterddr kshitij12345 saketh-are jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jgong5 commented:
> what's the procedure to land this into ideep?

No, I don't think we can do that since oneDNN expects the weight dim is the same as input. See this: https://oneapi-src.github.io/oneDNN/v2/dev_guide_prelu.html

>  I think the code would break in the case where input is 1D and weight is 0D, then the code dim_w[1] = weight.get_dim(0); would be executed and would break.

We can fix this one for the case when weight is a scalar. But it would be in the ATen/native/mkldnn/Prelu.cpp not in ideep, I guess, since ideep doesn't support 0-dim tensor. @yanbing-j

lezcano commented:
If ideep does not support 0-dim tensors, then the code there was correct all along.

About the ideep fix, the thing is that ideep does some manual resizing that's no longer necessary after this PR. That code can be removed. Note that oneDNN would get the tensors with the same number of dimensions, so nothing would change there, the only thing that would change is that the resizing would now be done at the ATen level, rather than at the ideep level.

lezcano commented:
Can I please get reviews from:
@kulinseth for the MPS changes
@albanD @soulitzer for the changes in the derivatives
@jbschlosser @albanD for the PR in general

I think that the quantisation and the MKLDNN changes are minimal so they don't need much of a review, or they can be reviewed by non-domain experts.

jbschlosser commented on aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:
```diff
@@ -680,14 +680,18 @@ static void qprelu_out_kernel(Tensor& out,
-  if (input_ndim > 1) {
-    sizes[1] = qw.numel();
+  // This logic is present in at::prelu and repeated here, as this path can be
+  // hit via quantizied::prelu, which is registered under quantized/cpu/qprelu.cpu
```
```suggestion
  // hit via quantized::prelu, which is registered under quantized/cpu/qprelu.cpu
```

lezcano committed:
Update on "Implement PReLU in a compositional way"


The PReLU implementation was all over the place. This lead to a number
of bugs like https://github.com/pytorch/pytorch/issues/68760.  We fix it by:
- Keeping the weird broadcasting logic it has as a CompositeImplicit kernel that calls into a second kernel
- This second kernel is just a good-ol' pointwise kernel.
- We implement the derivative for the pointwise kernel via TI as well for speed.
- We implement the second derivative for the pointwise kernel and the forward AD derivatives compositionally

This fixes a number of issues:
- We don't perform copies any more when the inputs are not contiguous
- The derivatives are now correct
- We fix vmap and many other functorch-related issues.
- CPU and CUDA now share the relevant broadcasting logic
- The implementation is about 1/3 the length.


Fixes https://github.com/pytorch/pytorch/issues/68760
Fixes https://github.com/pytorch/pytorch/issues/89895

cc albanD mruberry jbschlosser walterddr kshitij12345 saketh-are jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

albanD commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -4654,22 +4654,21 @@
 
 - func: prelu(Tensor self, Tensor weight) -> Tensor
   variants: function, method
+  autogen: prelu.out
```
Do you actually need a prelu.out since this is CompositeImplicit ?

lezcano replied:
Leaving it as people may still want to call `torch.prelu(x, out=out)`

jbschlosser reviewed (approved):
Thanks for all your work getting this cleaned up! I'm good with it on my end though I think ideally we would avoid adding any ops.

Without adding `_prelu_kernel` / `_prelu_backward_kernel`, I think we could still get de-duplication of the broadcasting logic by dispatching CPU / CUDA to the same function and using `REGISTER_DISPATCH` dispatching from there. However, I do understand that this doesn't fix vmap / simplify writing batch rules or simplify derivative computation, which I think are valuable. So imo the tradeoffs are worth it.

albanD commented on tools/autograd/derivatives.yaml:
```diff
@@ -1987,14 +1987,20 @@
+                (grads[1].defined() ? at::where(self >= 0, grads[0], grads[0] * weight + grads[1] * self)
+                                    : at::where(self >= 0, grads[0], grads[0] * weight))
+                                    : at::where(self >= 0, at::zeros({}, grad_output.options()), grads[1] * self)"
+
```
extra newline

albanD reviewed (approved):
SGTM

lezcano committed:
Update on "Implement PReLU in a compositional way"


The PReLU implementation was all over the place. This lead to a number
of bugs like https://github.com/pytorch/pytorch/issues/68760.  We fix it by:
- Keeping the weird broadcasting logic it has as a CompositeImplicit kernel that calls into a second kernel
- This second kernel is just a good-ol' pointwise kernel.
- We implement the derivative for the pointwise kernel via TI as well for speed.
- We implement the second derivative for the pointwise kernel and the forward AD derivatives compositionally

This fixes a number of issues:
- We don't perform copies any more when the inputs are not contiguous
- The derivatives are now correct
- We fix vmap and many other functorch-related issues.
- CPU and CUDA now share the relevant broadcasting logic
- The implementation is about 1/3 the length.


Fixes https://github.com/pytorch/pytorch/issues/68760
Fixes https://github.com/pytorch/pytorch/issues/89895

cc albanD mruberry jbschlosser walterddr kshitij12345 saketh-are jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

lezcano commented:
Great coordinated reviews. Even better being the two of them approvals :D

@pytorchbot merge

yanbing-j commented:
> We can fix this one for the case when weight is a scalar. But it would be in the ATen/native/mkldnn/Prelu.cpp not in ideep, I guess, since ideep doesn't support 0-dim tensor. @yanbing-j

Yes. We have changed the weight dimension to input dimension in https://github.com/pytorch/pytorch/pull/91238/files#diff-3dc8af925ec5d52a7cc5722279de60aeac5458097166318bcf6bfd58c420e629L710. The scenario of 1D input and 0D weight is supported. @jgong5 @lezcano 

> the only thing that would change is that the resizing would now be done at the ATen level, rather than at the ideep level.

Correct. The resizing has been done in ATen level, if the dimensions of src and weight are the same, the resizing in ideep level will not be invoked, and it can be regarded as a double check. @lezcano

Xia-Weiwen commented:
Hi @lezcano Looks like the quantization issue is fixed, right? https://github.com/pytorch/pytorch/pull/91238#issuecomment-1366882486

lezcano commented:
yeah, all good on the quantisation end. Thank you for checking in though @Xia-Weiwen

lezcano commented:
@pytorchbot merge -f "flaky tests"
